### ISSUE TITLE
[Jepsen] [YSQL] Postgres processes leftover when using LXC nodes

### ISSUE THREAD

**Issue by frozenspider** [yugabyte/yugabyte-db#4577]
Jira Link: [DB-2487](https://yugabyte.atlassian.net/browse/DB-2487)
This is very similar to #2075, but is only reproducible when using LXC nodes instead of real nodes.
Using the same steps as described in #2075 (except not using real nodes), logging number of postgres processes via `ps aux | grep postgres | wc -l` hourly using cron job, and killing all with `sudo pkill -9 postgres` before starting Jepsen run, I observed number of `postgres` processes on a host node rising from 30 to 800, with number of per-LXC `postgres` processes increasing from 7 to 180.
Most of those processes were stuck in `authentication` state, and some - in `SELECT` or `INSERT`. This is also different to the actual issue reported in #2075

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-4577-analysis.md`):

# Analysis: yugabyte/yugabyte-db#4577 (DB-2487)

Issue: "[Jepsen] [YSQL] Postgres processes leftover when using LXC nodes" (filed 2020-05-26 against the
2.1.8-era code, open, no comments). Reporter ran the #2075 Jepsen YSQL workload on LXC nodes and saw the
per-node count of `postgres` processes grow (7 -> 180 per LXC, 30 -> 800 on the host) across a run, most
of them parked in the `authentication` ps state and some in `SELECT`/`INSERT`; `pkill -9 postgres` was
needed before each run. Component: YSQL (`src/postgres/`, `src/yb/yql/pgwrapper/`, `src/yb/yql/pggate/`).
Kind: process-lifecycle leak (hang class). Related: #2075 (same symptom on 1.3.1, fixed by 73e966e5 / #1857).

## Verdict
NOT-REPRODUCED -- a faithful single-container lifecycle probe (tserver freeze, postmaster SIGINT/SIGKILL,
tserver SIGKILL, tserver restart; 7 phases, each with a clean-return control) leaves zero residual backends
on every required target for every path the report depends on (startup-phase parking, postmaster death,
tserver death/restart). The only residue observed anywhere is on the master *preview* image 2.25.2.0-b359
in a path the report does not exercise (an external SIGINT fast shutdown while the tserver is unresponsive,
phase B); that image predates #26758, which master source already contains (`postmaster.c:2021-2032`).
Reporter-environment differences: Jepsen + LXC harness, RF3 with iptables partitions and `kill -9` nemeses
over hours, and 2.1.8-era code lacking every hardening listed below.

## Root cause
Reported-era mechanism (source at tag `v2.1.8.1`, commit 16b0e59456), four cooperating defects:

1. Unbounded backend startup wait. `YBInitPostgresBackend` ran before `PerformAuthentication`
   (`src/postgres/src/backend/utils/init/postinit.c:675-677` vs `:769`), so a backend whose pggate init
   could not reach the cluster sat under the PG11 initial ps label `authentication`
   (`src/postgres/src/backend/postmaster/postmaster.c:4320-4321`, `init_ps_display(..., "authentication")`).
   pggate's client init retried `Build()` forever with a 1s sleep (`src/yb/client/async_initializer.cc:81-97`),
   so a client that gave up left a backend parked indefinitely. Each Jepsen reconnect against a partitioned
   or restarting node added one.
2. Blocked backends ignored fast shutdown. `die()` only set `InterruptPending`/`ProcDiePending` and the
   latch (`src/postgres/src/backend/tcop/postgres.c:2725-2730`); a backend blocked inside a C++ RPC/future
   wait never returned to `ProcessInterrupts`. The postmaster escalated to SIGKILL only for immediate
   shutdown or crash recovery (`postmaster.c:1843-1845`), never for the fast shutdown that the tserver's
   death delivered (`src/yb/yql/pgwrapper/pg_wrapper.cc:303`, `SetParentDeathSignal(SIGINT)`), so the old
   postmaster hung waiting on its stuck children.
3. Tserver restart orphaned them. `PgSupervisor::CleanupOldServerUnlocked` (`pg_wrapper.cc:485-503`)
   SIGKILLed the old postmaster immediately on the next tserver start; its still-blocked children were
   re-parented to init.
4. No parent-death signal for backends. Neither `postmaster.c` nor `pmsignal.c` at `v2.1.8.1` set
   `PR_SET_PDEATHSIG` for backends (PG11 has no `PostmasterDeathSignalInit`), and a backend blocked in C++
   never polls `PostmasterIsAlive()`, so orphans from (3) lived until manually killed. Their ps states are
   exactly the report's: `authentication` (parked in (1)) and `SELECT`/`INSERT` (blocked DML from (2)).

Why LXC-only in 2020 is not recoverable from the report; the mechanism itself is platform-independent
(the LXC setup changes only how often nemeses left nodes unreachable while clients kept reconnecting).

Current source (local master 46a9f7adfd) closes each defect; every one was verified in the runtime matrix:

- (1) bounded: the create-session heartbeat RPC carries a `pg_client_heartbeat_interval_ms - 1s` = 9s
  timeout (`src/yb/yql/pggate/pg_client.cc:2047-2052`) and `PgClient::Impl::Start` fails the backend when it
  expires (`pg_client.cc:726-737`); the parked label is now `initializing` on PG15 builds
  (`postmaster.c:5075`; `postinit.c:1131` `YBInitPostgresBackend` still precedes `postinit.c:1312`
  `PerformAuthentication`, `postinit.c:277` sets `authentication` only afterwards). Observed: parked
  backends exit at 9s with `FATAL: Heartbeat RPC (request call id 1) to 127.0.0.1:9100 timed out after 9.000s`.
- (2) `die()` calls `YBCInterruptPgGate()` (`postgres.c:3504-3534`, `:3516`) which shuts the messenger down
  from a helper thread (`src/yb/yql/pggate/pggate.cc:564-606`) so RPC waits abort (#11343, c28ce4c7f3); the
  postmaster SIGKILLs children still alive 5s into a *fast* shutdown (`postmaster.c:2018-2032`, #26758,
  0020131278).
- (3) `PgWrapper::CleanupLockFileAndKillHungPg` (`pg_wrapper.cc:1471-1567`) only kills a still-running
  postmaster after a 6s grace (`:1543`) and treats a dead pid as a stale lock (`:1532-1535`, #24396 111b65d1f1
  + #26758).
- (4) decisive: every backend requests SIGKILL on postmaster death -- PG15 lineage
  `src/postgres/src/backend/storage/ipc/pmsignal.c:443-448` in `PostmasterDeathSignalInit` (called from
  `src/postgres/src/backend/utils/init/miscinit.c:172`); PG11 lineage (2024.2) `postmaster.c:4501-4508`
  in the `BackendStartup` fork child (#17805/#17965, 642980eef8; moved for PG15 by 632949c955). An orphan
  in state (1) or (2) is therefore impossible regardless of what it is blocked on.

Still affected: none of the required targets for the reported paths. The 2.25.2.0-b359 preview lacks
the #26758 escalation (its `postmaster.c` clause is the upstream `Shutdown >= ImmediateShutdown` form) and
its tserver uses SIGQUIT for PG shutdown (`src/yb/yql/process_wrapper/process_wrapper.cc:42`,
`pg_wrapper.cc:890`), so only an externally sent SIGINT reaches the gap. No master fix PR is warranted.

## Trigger matrix
Predictions derived from the source above before the runs; observed = probe v2 on the five images.

| phase / condition | predicted | observed (2024.2 / 2025.1 / 2025.2 / 2026.1 / master-preview 2.25.2.0) |
|---|---|---|
| control: 6 idle sessions open, then clients close | 6 then 0 | 6 -> 0 within 1s on all five |
| A: tserver SIGSTOP; 6 connects with 3s client timeout | backends park under `authentication` (PG11) / `initializing` (PG15); exit at ~9s (heartbeat bound) while still frozen | label as predicted; exit at 9s / 9s / 9s / 9s / 15s after connect, tserver still frozen; FATAL `Heartbeat RPC ... timed out after 9.000s` on the four release targets |
| B: 6 sessions blocked in SELECT/INSERT on frozen tserver; SIGINT postmaster | 0 within <=5s (interrupt path; 5s SIGKILL escalation); 2.25.2.0 may stall (no escalation) | 5s / 0s / 1s / 1s / **stall**: 6 blocked backends + postmaster persist >=150s while frozen (supplementary run), 5 of 6 still attached >=100s after SIGCONT; YSQL unavailable until the postmaster dies |
| C: SIGKILL postmaster; 3 idle + 3 idle-in-transaction | 0 orphans at +1s (backend PDEATHSIG=SIGKILL) | 0 / 0 / 0 / 0 / 0 at +1s; supervisor restart ok on all |
| E: SIGKILL postmaster while 6 backends parked in startup phase, tserver frozen | 0 orphans at +1s | 0 / 0 / 0 / 0 / 0 at +1s and +3s; ysql ok after SIGCONT; stale `postmaster.pid`/socket lock removed by the tserver on restart |
| D: SIGKILL healthy tserver; 4 idle + 2 active backends | whole postgres tree gone <=6s | 0 processes at +1s on all five; yugabyted restarts the tserver within ~3s; new postmaster serves |
| F: SIGKILL frozen tserver; 3 idle + 3 blocked backends | gone <=6s (escalation); 2.25.2.0 immediate (SIGQUIT) | 2024.2: 3 blocked remain at +1s, all gone at +5s; 2025.1: 3 remain at +1s, SIGKILLed by escalation (`recalcitrant` log) by +5s; 2025.2 / 2026.1 / 2.25.2.0: 0 at +1s |

Every phase ended with `probe_backends=0` and a healthy YSQL on every image (final line of each cell log);
no process ever survived its postmaster's death; no process ever survived past a bounded timeout except in
the 2.25.2.0 phase-B stall, where the residue stayed attached to a live postmaster (ppid = postmaster) and
no new backends could accumulate (the postmaster rejects connections during shutdown).

## Reproduction
Files (all under `/tmp/claude/yb4577/`): `yb4577-probe.txt` (in-container probe, sha256
7f82e2ef3b7ef6905640e23311113829e87e362571d95ab168467a488e2cae4b), `yb4577-run-matrix.txt` (host driver),
`yb4577-probe-b-long.txt` (supplementary 150s phase B), `yb4577-cell-<label>.txt` (raw logs),
`yb4577-matrix-summary.txt` (RESULT lines per target), `yb4577-dockerhub-freshness.txt`.

```sh
# one target (aarch64 host; use the -x86_64 tags on amd64); ~2.5 min per image
bash /tmp/claude/yb4577/yb4577-run-matrix.txt 2026.1 2026.1.1.1-b2-aarch64
bash /tmp/claude/yb4577/yb4577-run-matrix.txt 2024.2 2024.2.11.0-b36-aarch64
bash /tmp/claude/yb4577/yb4577-run-matrix.txt 2025.1 2025.1.4.0-b103-aarch64
bash /tmp/claude/yb4577/yb4577-run-matrix.txt 2025.2 2025.2.6.0-b111-aarch64
bash /tmp/claude/yb4577/yb4577-run-matrix.txt master 2.25.2.0-b359-aarch64
```

The driver starts `podman run -d --init ... sleep infinity`, runs `yugabyted start --base_dir=/tmp/ybd
--advertise_address=127.0.0.1`, copies the probe in and runs it. The probe creates role `yb4577` so its
backends are identified exactly from the ps title (`postgres: yb4577 yugabyte 127.0.0.1(port) <state>`),
then executes control, A, B, C, E, D, F as described in the trigger matrix, printing `RESULT <phase> k=v`
lines. Signals used: `kill -STOP/-CONT <tserver>`, `kill -INT <postmaster>`, `kill -9 <postmaster>`,
`kill -9 <tserver>`. Failure signature to look for (would indicate the reported defect): any
`RESULT <phase> ... orphans_at_1s=N` with N>0 whose `orphan_ppids` is not the postmaster, or
`startup_backends_exit_seconds_after_connect=timeout(...)`, or `final probe_backends` > 0. Control: the
`control` phase and each phase's `ysql_after*=ok` / `leftover*=0` lines.

Adjustments per version: none required (same script on all five images). Observed image-specific
behaviour is recorded in the matrix (PG11 label `authentication`, 2.25.2.0 15s startup bound and phase-B stall).

## Versions tested
Resolved at: 2026-09-05T19:07Z
Resolution evidence: Docker Hub `https://hub.docker.com/v2/repositories/yugabytedb/yugabyte/tags?name=<branch>&ordering=last_updated`
listings (controller resolution; independently re-queried 2026-09-05T20:21:36Z with identical results, saved
in `/tmp/claude/yb4577/yb4577-dockerhub-freshness.txt`) plus GitHub tag -> commit lookups; `name=2.27.`,
`name=2.29.`, `name=2026.2` return count 0 (no newer preview line than 2.25.x).

| target | latest resolved version | source revision | observed repro/control | effective status |
|---|---|---|---|---|
| 2024.2 | 2024.2.11.0-b36 (docker.io/yugabytedb/yugabyte:2024.2.11.0-b36-aarch64, sha256:53c01c56315e245cc901cbc46e93f30fedb1f87833c7b1aa09d20fbeda27a288; amd64 sha256:6bb00eeb3dd97b93f36716080d2cfe96d1e61f864201d9a36ae6e3aeacb6d9ce; newest `2024.2` tag pushed 2026-08-25T16:02:12Z, prior 2024.2.10.0-b62; tag commit `[BACKPORT 2024.2][DEVOPS-3852] ...`) | v2024.2.11.0 = 04544fd94fb4e50d05994784f444b075a4a5dfdb | A: 6 parked as `authentication`, exit at 9s (heartbeat timeout FATAL); B: 0 at +5s; C/E: 0 orphans at +1s; D: 0 procs at +1s; F: 3 blocked at +1s, 0 at +5s; controls ok; final probe_backends=0 | fixed (PG11-lineage hardening `postmaster.c:4501-4508` + #26758 escalation + 9s bound) |
| 2025.1 | 2025.1.4.0-b103 (…:2025.1.4.0-b103-aarch64, sha256:08f1be0d6e074e0286e70067c12933a8cf536876d50fc43c52b1127d447ecdc0; amd64 sha256:fa5555536e8fe7d9522fb6c9c55f3c6062549f318b79c5902726097549263771; newest pushed 2026-04-20T15:59:42Z, prior 2025.1.3.2-b1; `[BACKPORT 2025.1][#30772] ...`) | v2025.1.4.0 = 690ac4141e57d060fd834312f396643e8c41f260 | A: `initializing`, exit at 9s; B: 0 at +0s; C/E: 0 orphans; D: 0 at +1s; F: 3 blocked at +1s, SIGKILLed by escalation (`recalcitrant` log) by +5s; controls ok; final 0 | fixed |
| 2025.2 | 2025.2.6.0-b111 (…:2025.2.6.0-b111-aarch64, sha256:4a1f7a94d3b70c61ecaf5e049521d691b16c9bf204ed01bdb6ab75d5beeb623e; amd64 sha256:5042b5090b49ef49959a4db9d05c90f5da6482aa72265d6ab4a23f8dc9ec0297; newest pushed 2026-09-04T22:23:02Z, prior 2025.2.5.2-b5; `[BACKPORT 2025.2.6][#33558] ...`) | v2025.2.6.0 = ae22ef35ff65de63303467d95ac3e6d2382feba9 | A: `initializing`, exit at 9s; B: 0 at +1s; C/E: 0 orphans; D/F: 0 at +1s; controls ok; final 0 | fixed |
| 2026.1 | 2026.1.1.1-b2 (= `latest`, manifest sha256:3926eedf0ff4b99426fbff533128ff16635ed5697a5cf086dde4fafec28e6f75; …:2026.1.1.1-b2-aarch64 sha256:d27eede72bbc669e348e9920318589b4bb68e98204a241ec6bfaba04bf4f6699; amd64 sha256:36741c614b0a28498a3447a01b952240905a0ad87c9f68dbcbb20e65bc3d2e9f; newest pushed 2026-08-18T20:03:16Z, prior 2026.1.1.0-b91; `[BACKPORT 2026.1.1][PLAT-21546] ...`) | v2026.1.1.1 = 3b010e021f6027744bf9e016b17e108a2eb24e44 | A: `initializing`, exit at 9s; B: 0 at +1s; C/E: 0 orphans; D/F: 0 at +1s; controls ok; final 0 | fixed |
| master | preview 2.25.2.0-b359 (…:2.25.2.0-b359-aarch64, sha256:3f9cd1b8f520cc78e82ace51174be73343a551148bc6c9a9f821a22f1d2858c5; amd64 sha256:e69ddcd248db7ee5c097d8c955d3657cf7fd9c21743a4589281767487ab5832b; Docker Hub `name=2.25.` newest pushed 2025-05-20T18:34:42Z; `name=2.27.`, `name=2.29.`, `name=2026.2` count 0) | v2.25.2.0 = 03bbc9f31eae5c3fecabd8f0afa5e99211e6080d; master HEAD b8477c4284908e7ff9314fb2b31014d12f20cc0c (2026-09-04T23:18:07Z) | A: `initializing`, exit at 15s; C/E: 0 orphans at +1s; D/F: 0 at +1s (SIGQUIT PDEATHSIG); controls ok; final 0. B: stall -- 6 blocked backends + postmaster persist >=150s while tserver frozen, 5 remain >=100s after SIGCONT (supplementary run) | fixed for the reported paths (A/C/D/E/F); the phase-B fast-shutdown gap is closed in master source by #26758 (`postmaster.c:2021-2032` at 46a9f7adfd), not runtime-verifiable on a published image |

First published release containing each hardening (tag/commit ancestry, provenance only -- effectiveness
is the runtime matrix above):

- Backend `PR_SET_PDEATHSIG=SIGKILL` (#17805/#17965, master 642980eef8 2023-07-19): PG11 lineage first in
  v2.19.2.0; backports v2.14.12.0, v2.16.7.0, v2.18.2.1; all v2.20.x/v2.21.x/v2.23.x incl. v2024.2.11.0
  (`postmaster.c:4508`). PG15 lineage via 632949c955 (2024-07-29) in `pmsignal.c`: v2.25.0.0+, v2025.1.x,
  v2025.2.x, v2026.1.x (verified by `git show <tag>:...` for v2.1.8.1..v2.25.2.0 and GitHub contents for v2025.2.6.0).
- `die()` -> `YBCInterruptPgGate()` (#11343, c28ce4c7f3 2022-06-16): present at v2024.2.11.0 (`postgres.c:2867`),
  v2025.1.4.0 (`:3246`), v2025.2.6.0, v2026.1.1.1 (`:3515`), v2.25.2.0 (`:3144`).
- Bounded create-session wait (`heartbeat_controller_.set_timeout(... - 1s)`): present at v2024.2.11.0
  (`pg_client.cc:1419`), v2025.1.4.0 (`:1810`), v2025.2.6.0 (`:1963`), v2026.1.1.1 (`:1994`), v2.25.2.0 (`:1465`);
  the unbounded `AsyncClientInitialiser` path left pggate with #9936 (fbcb98e22f, 2022-03-11).
- Fast-shutdown SIGKILL escalation + 6s cleanup grace + SIGINT PDEATHSIG (#26758, master 0020131278
  2025-06-06): v2024.2.11.0 (`postmaster.c:1898`, `pg_wrapper.cc:1213`), v2025.1.4.0 (`:2008`, `:1378`),
  v2025.2.6.0 (`:2015`, `:1404`), v2026.1.1.1 (`:2022`, `:1449`); absent from v2.25.2.0.
- Per-start lock-file cleanup / kill of a hung old postmaster (#24396, 111b65d1f1 2024-10-23):
  `CleanupLockFileAndKillHungPg` present at all five tags (v2.25.2.0 without the 6s grace).

## Proposed fix direction
No code change is proposed for the reported defect: all four contributing mechanisms are closed on every
in-support branch and on master source, and the runtime matrix shows cleanup on every target. Recommend
closing #4577 as resolved by the hardenings above (with #2075 as the earlier instance of the same symptom).
If the maintainers want the phase-B fast-shutdown stall observed on the 2.25.2.0 preview covered for
future previews, it is already addressed by #26758 (`postmaster.c:2018-2032`); no additional change.

## Test strategy
Existing coverage of the hardened paths: `PgPostmasterExitTest.SignalBackendOnPostmasterDeath` and
`SignalIdleBackendOnPostmasterDeath` (`src/yb/yql/pgwrapper/pg_libpq-test.cc`; backend PDEATHSIG),
`PgWrapperTest.TestHungPg` and `PgWrapperOneNodeClusterTest.TestPostgresLockFiles`
(`src/yb/yql/pgwrapper/pg_wrapper-test.cc`; fast-shutdown escalation and lock-file cleanup),
`PgMiniTest.NoWaitForRPCOnTermination` (`src/yb/yql/pgwrapper/pg_mini-test.cc:2491`; #11343). Public-entry-point check for a maintainer: the probe in
this artifact (official image, `yugabyted`, `ysqlsh`, `kill`) -- e.g.
`./yb_build.sh release --cxx-test pgwrapper_pg_libpq-test --gtest_filter PgPostmasterExitTest.SignalBackendOnPostmasterDeath`.

## Confidentiality check
Contents are public: GitHub issue text, public commit SHAs/tags, official Docker Hub tags/digests,
loopback addresses, synthetic role/table names (`yb4577`, `t`). No customer identifiers, PII, internal
hostnames/URLs, or secrets in this artifact, the probe scripts, or the cell logs.


[DEVOPS-3852]: https://yugabyte.atlassian.net/browse/DEVOPS-3852?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4577-probe.txt`):

#!/bin/bash
# yb4577 probe v2 -- YSQL backend lifecycle under tserver freeze / postmaster and tserver termination.
# Target: yugabyte/yugabyte-db#4577 ("postgres processes leftover", most parked in the startup phase,
# some in SELECT/INSERT). Runs INSIDE an official yugabytedb/yugabyte container after
# `yugabyted start --base_dir=/tmp/ybd`. Single node, RF1, trust auth.
#
# Every probe-owned session connects as role yb4577 so its backend can be identified exactly from the
# ps title ("postgres: [cluster: X ]yb4577 yugabyte 127.0.0.1(port) <state>"). Pre-existing backends
# (yugabyted's own poller etc.) are therefore never miscounted. Every phase prints RESULT lines.
#
# Phases (each with an in-phase control = clean return of the probe backends to zero):
#   control : 6 idle probe sessions appear, then vanish when their clients close.
#   A       : tserver frozen (SIGSTOP); 6 connection attempts with a 3s client timeout. Do backends park
#             in the startup phase (label per build), and for how long while the tserver stays frozen?
#   B       : 6 established sessions blocked in SELECT/INSERT on a frozen tserver; SIGINT to postmaster
#             (fast shutdown, the signal the tserver's death delivers on SIGINT-PDEATHSIG builds).
#   C       : SIGKILL the postmaster with 3 idle + 3 idle-in-transaction backends.
#   E       : SIGKILL the postmaster while 6 backends are parked in the startup phase (frozen tserver).
#   D       : SIGKILL a healthy tserver with 4 idle + 2 active backends; then relaunch the tserver.
#   F       : SIGKILL a frozen tserver while 3 idle + 3 blocked backends exist; then relaunch the tserver.
set -u
BIN=/home/yugabyte/bin
BASE=/tmp/ybd
ADMIN="$BIN/ysqlsh -h 127.0.0.1 -p 5433 -U yugabyte -d yugabyte -X -q -A -t"
PSQL="$BIN/ysqlsh -h 127.0.0.1 -p 5433 -U yb4577 -d yugabyte -X -q -A -t"
W=/tmp/yb4577
mkdir -p "$W"

ts() { date -u +%H:%M:%S; }
log() { echo "$(ts) $*"; }
res() { echo "RESULT $*"; }

PS() { ps -ewwo pid=,ppid=,stat=,args=; }
# Every live (non-zombie) postgres process: postmaster (".../postgres -D"), aux processes and backends.
pgprocs() { PS | awk '($4 ~ /\/postgres$/ || $4=="postgres:") && $3 !~ /^Z/'; }
npgprocs() { pgprocs | wc -l | tr -d ' '; }
# Probe-owned client backends only.
probe_backends() { PS | awk '$4=="postgres:" && $3 !~ /^Z/ && $0 ~ / yb4577 yugabyte 127\.0\.0\.1\([0-9]+\)/'; }
nprobe() { probe_backends | wc -l | tr -d ' '; }
probe_states() { probe_backends | sed -E 's/^.* yb4577 yugabyte 127\.0\.0\.1\([0-9]+\) ?//' | sed 's/^$/(blank)/' | sort | uniq -c | awk '{c=$1; $1=""; sub(/^ /,""); printf "%s=%s;", $0, c}'; }
probe_ppids() { probe_backends | awk '{print $2}' | sort -u | tr '\n' ',' ; }
# Probe backends still in the startup phase (label depends on the PG major: PG11 "authentication",
# PG15 "initializing" until auth, then "startup"; blank title = before ps title is set).
nstartup() { probe_backends | sed -E 's/^.* yb4577 yugabyte 127\.0\.0\.1\([0-9]+\) ?//' | grep -c -E '^(initializing|authentication|startup|)$'; }
# Live (non-zombie) yb-tserver; a SIGKILLed tserver lingers as a zombie until yugabyted reaps it.
tserver_pid() { ps -eo pid=,stat=,comm= | awk '$3=="yb-tserver" && $2 !~ /^Z/ {print $1}' | tail -1; }
postmaster_pid() { head -1 "$BASE"/data/pg_data/postmaster.pid 2>/dev/null; }
postmaster_alive() { local p; p=$(postmaster_pid); [ -n "$p" ] && kill -0 "$p" 2>/dev/null && grep -q '/postgres' "/proc/$p/cmdline" 2>/dev/null; }
ysql_ok() { timeout 8 $PSQL -c 'select 1' >/dev/null 2>&1; }
wait_ysql() { local i; for i in $(seq 1 "$1"); do ysql_ok && return 0; sleep 1; done; return 1; }
pglog_grep() { find "$BASE" \( -name 'postgresql-*.log' -o -name '*.err' \) 2>/dev/null | xargs -r grep -h -E "$1" | tail -n "${2:-3}"; }
tslog_grep() { { cat "$W"/tserver-relaunch*.log 2>/dev/null; find "$BASE" \( -name 'yb-tserver*INFO*' -o -name 'yb-tserver*WARNING*' -o -name 'tserver.err' \) 2>/dev/null | xargs -r cat; } | grep -h -E "$1" | tail -n "${2:-4}"; }
now_s() { date +%s; }

sample() { log "$1: $2 probe_backends=$(nprobe) [$(probe_states)] ppids=[$(probe_ppids)] pgprocs=$(npgprocs)"; }
# wait until no probe backend is left; prints elapsed seconds or "timeout(last=N)".
wait_zero() { local i n; for i in $(seq 0 "$1"); do n=$(nprobe); [ "$n" -eq 0 ] && { echo "$i"; return 0; }; sleep 1; done; echo "timeout(last=$(nprobe))"; return 1; }
# wait until no probe backend is in the startup phase; prints elapsed seconds or "timeout(last=N)".
wait_startup_zero() { local i n; for i in $(seq 0 "$1"); do n=$(nstartup); [ "$n" -eq 0 ] && { echo "$i"; return 0; }; sleep 1; done; echo "timeout(last=$(nstartup))"; return 1; }
# wait until at least N probe backends exist; prints elapsed seconds or "timeout(last=N)".
wait_atleast() { local i n; for i in $(seq 0 "$2"); do n=$(nprobe); [ "$n" -ge "$1" ] && { echo "$i"; return 0; }; sleep 1; done; echo "timeout(last=$(nprobe))"; return 1; }
wait_pgprocs_zero() { local i n; for i in $(seq 0 "$1"); do n=$(npgprocs); [ "$n" -eq 0 ] && { echo "$i"; return 0; }; sleep 1; done; echo "timeout(last=$(npgprocs))"; return 1; }

# Sessions held open by a pipe: stdin is fed by `sleep <tag>`; killing that sleep closes the client.
idle_client() { # $1=sleep tag (unique per group)  $2=optional preamble SQL (e.g. "BEGIN;")
  ( { [ -n "${2:-}" ] && echo "$2"; exec sleep "$1"; } | $PSQL >"$W/idle-$1-$BASHPID.out" 2>&1 ) &
}
# The image ships single-binary coreutils: the sleeper's cmdline is
# "/usr/bin/coreutils --coreutils-prog-shebang=sleep /usr/bin/sleep <tag>", so anchor at the end only.
close_clients() { pkill -f "/sleep $1\$" 2>/dev/null; }
# Query that first runs client-side-visible pg_sleep(3) (no tserver RPC) and then hits DocDB.
blocked_client() { # $1=SELECT|INSERT  $2=id
  if [ "$1" = SELECT ]; then
    ( $PSQL -c "SELECT pg_sleep(3); SELECT count(*) FROM t;" >"$W/blocked-$2.out" 2>&1 ) &
  else
    ( $PSQL -c "SELECT pg_sleep(3); INSERT INTO t VALUES (1000 + $2, $2);" >"$W/blocked-$2.out" 2>&1 ) &
  fi
}
timeout_client() { # $1=id ; libpq gives up after 3s, backend keeps whatever state it is in
  ( PGCONNECT_TIMEOUT=3 $PSQL -c 'select 1' >"$W/tmo-$1.out" 2>&1 ) &
}
relaunch_tserver() { # $1=phase $2=old pid ; yugabyted normally restarts the tserver itself within ~2s
  local i p
  for i in 1 2 3 4 5; do p=$(tserver_pid); [ -n "$p" ] && [ "$p" != "$2" ] && { log "$1: tserver restarted by yugabyted (new pid $p)"; res "$1" tserver_restarted_by=yugabyted new_tserver="$p"; return; }; sleep 1; done
  log "$1: relaunching tserver with the captured command line"
  ( cd / && nohup bash -c "exec $TS_CMDLINE" >"$W/tserver-relaunch-$1.log" 2>&1 & )
  sleep 2; log "$1: relaunched tserver pid=$(tserver_pid)"; res "$1" tserver_restarted_by=probe new_tserver="$(tserver_pid)"
}

log "== yb4577 probe v2 start; version: $($BIN/yb-tserver --version 2>&1 | head -1)"
log "kernel: $(uname -r); ps: $(ps --version 2>&1 | head -1)"
TS=$(tserver_pid); PM=$(postmaster_pid)
TS_CMDLINE=$(tr '\0' ' ' <"/proc/$TS/cmdline")
log "tserver pid=$TS postmaster pid=$PM (ppid of postmaster: $(awk '{print $4}' /proc/$PM/stat))"
log "tserver cmdline: $TS_CMDLINE"
$ADMIN -c "DROP ROLE IF EXISTS yb4577; CREATE ROLE yb4577 LOGIN SUPERUSER;" >/dev/null 2>&1
$PSQL -c 'DROP TABLE IF EXISTS t; CREATE TABLE t (k int PRIMARY KEY, v int); INSERT INTO t SELECT i, i FROM generate_series(1,100) i;' >/dev/null 2>&1
log "pre-existing (non-probe) client backends: $(PS | awk '$4=="postgres:" && $0 ~ /\([0-9]+\)/ && $3 !~ /^Z/' | sed -E 's/^ *[0-9]+ +[0-9]+ +[^ ]+ +//' | tr '\n' '|')"
log "all postgres procs at start: $(npgprocs); probe backends: $(nprobe)"
res baseline pgprocs=$(npgprocs) probe_backends=$(nprobe)

# ---- control -----------------------------------------------------------------------------------
for i in 1 2 3 4 5 6; do idle_client 3600; done
t=$(wait_atleast 6 20); sample control "6 idle sessions opened (+${t}s)"
res control connected=$(nprobe) states="$(probe_states)"
close_clients 3600; t=$(wait_zero 20); sample control "clients closed (+${t}s)"
res control seconds_to_zero_after_client_close=$t leftover=$(nprobe)

# ---- A: frozen tserver, 6 timed-out connection attempts ---------------------------------------
TS=$(tserver_pid); kill -STOP "$TS"; log "A: SIGSTOP tserver $TS"
T0=$(now_s); for i in 1 2 3 4 5 6; do timeout_client $i; done
sleep 5; sample A "+5s (clients gave up at 3s)"
res A startup_phase_backends_at_5s=$(nstartup) label="$(probe_states)"
res A client_result="$(cat "$W"/tmo-*.out 2>/dev/null | sort | uniq -c | sed 's/^ *//' | tr '\n' '|')"
t=$(wait_startup_zero 60); sample A "startup-phase backends gone while tserver still frozen (+${t}s after the +5s sample)"
res A startup_backends_exit_seconds_after_connect=$([ "${t#timeout}" = "$t" ] && echo $(( $(now_s) - T0 )) || echo "$t") still_frozen=yes
t=$(wait_zero 10); res A all_probe_backends_zero_while_frozen=$([ "${t#timeout}" = "$t" ] && echo yes || echo "no($t)")
kill -CONT "$TS"; log "A: SIGCONT tserver"
t2=$(wait_zero 20); sample A "after SIGCONT (+${t2}s)"
res A leftover_after_cont=$(nprobe)
wait_ysql 60 && log "A: ysql ok" || log "A: ysql NOT ok"
res A ysql_after_cont=$(ysql_ok && echo ok || echo fail)
res A pglog_startup_exit_reason="$(pglog_grep 'FATAL.*(Heartbeat|imed out|ession|nitializ)' 2 | sed -E 's/^[^ ]+ [^ ]+ [^ ]+ //' | cut -c1-160 | tr '\n' '|')"
rm -f "$W"/tmo-*.out

# ---- B: sessions blocked in SELECT/INSERT on a frozen tserver; SIGINT to the postmaster ----------
for i in 1 2 3; do blocked_client SELECT $i; done
for i in 4 5 6; do blocked_client INSERT $i; done
t=$(wait_atleast 6 20); sample B "6 sessions connected, running pg_sleep(3) (+${t}s)"
TS=$(tserver_pid); kill -STOP "$TS"; log "B: SIGSTOP tserver $TS"
sleep 4; sample B "+4s: pg_sleep done, DocDB requests outstanding against the frozen tserver"
res B blocked_states="$(probe_states)" blocked_count=$(nprobe)
PM=$(postmaster_pid); kill -INT "$PM"; log "B: SIGINT postmaster $PM (fast shutdown)"
t=$(wait_zero 15); sample B "after SIGINT (+${t}s, tserver still frozen)"
res B seconds_to_zero_after_sigint=$t leftover=$(nprobe) pgprocs_left=$(npgprocs) postmaster_alive=$(postmaster_alive && echo yes || echo no)
res B recalcitrant_sigkill_log=$(pglog_grep 'recalcitrant' 5 | wc -l | tr -d ' ')
kill -CONT "$TS"; log "B: SIGCONT tserver"
wait_ysql 90 && log "B: ysql ok, postmaster pid=$(postmaster_pid)" || log "B: ysql NOT ok"
res B ysql_after_cont=$(ysql_ok && echo ok || echo fail) new_postmaster=$(postmaster_pid)
res B client_errors="$(cat "$W"/blocked-*.out 2>/dev/null | sort | uniq -c | sed 's/^ *//' | cut -c1-120 | tr '\n' '|')"
rm -f "$W"/blocked-*.out

# ---- C: SIGKILL the postmaster with idle + idle-in-transaction backends --------------------------
for i in 1 2 3; do idle_client 3601; done
for i in 4 5 6; do idle_client 3602 "BEGIN;"; done
t=$(wait_atleast 6 20); sleep 2; sample C "6 sessions (+${t}s)"
res C states_before="$(probe_states)"
PM=$(postmaster_pid); kill -9 "$PM"; log "C: SIGKILL postmaster $PM"
sleep 1; sample C "+1s after postmaster SIGKILL"
res C orphans_at_1s=$(nprobe) orphan_ppids="$(probe_ppids)" orphan_states="$(probe_states)"
t=$(wait_zero 12); sample C "clients still connected (+${t}s)"
res C seconds_to_zero_clients_connected=$t leftover_clients_connected=$(nprobe)
close_clients 3601; close_clients 3602
t=$(wait_zero 10); sample C "clients closed (+${t}s)"
res C leftover_after_client_close=$(nprobe) seconds_to_zero_after_client_close=$t
wait_ysql 90 && log "C: ysql ok, postmaster pid=$(postmaster_pid)" || log "C: ysql NOT ok"
res C ysql_after=$(ysql_ok && echo ok || echo fail) new_postmaster=$(postmaster_pid)
res C tserver_cleanup_log="$(tslog_grep 'older postgres process|stale postgres lock file|Restarting PostgreSQL' 3 | sed -E 's/^[^]]*\] //' | cut -c1-140 | tr '\n' '|')"

# ---- E: SIGKILL the postmaster while backends are parked in the startup phase (frozen tserver) ---
TS=$(tserver_pid); kill -STOP "$TS"; log "E: SIGSTOP tserver $TS"
for i in 1 2 3 4 5 6; do timeout_client $i; done
sleep 4; sample E "+4s: startup-phase backends"
res E startup_phase_backends_before=$(nstartup) label="$(probe_states)"
PM=$(postmaster_pid); kill -9 "$PM"; log "E: SIGKILL postmaster $PM"
sleep 1; sample E "+1s after postmaster SIGKILL (tserver frozen)"
res E orphans_at_1s=$(nprobe) orphan_ppids="$(probe_ppids)" orphan_states="$(probe_states)"
sleep 2; sample E "+3s after postmaster SIGKILL (tserver frozen)"
res E orphans_at_3s=$(nprobe)
t=$(wait_zero 30); sample E "waited for exit while tserver frozen (+${t}s after the +3s sample)"
res E seconds_to_zero_while_frozen_after_3s=$t
kill -CONT "$TS"; log "E: SIGCONT tserver"
t2=$(wait_zero 20); sample E "after SIGCONT (+${t2}s)"
res E leftover_after_cont=$(nprobe)
wait_ysql 90 && log "E: ysql ok, postmaster pid=$(postmaster_pid)" || log "E: ysql NOT ok"
res E ysql_after=$(ysql_ok && echo ok || echo fail) new_postmaster=$(postmaster_pid)
res E tserver_cleanup_log="$(tslog_grep 'older postgres process|stale postgres lock file|Restarting PostgreSQL' 3 | sed -E 's/^[^]]*\] //' | cut -c1-140 | tr '\n' '|')"
rm -f "$W"/tmo-*.out

# ---- D: SIGKILL a healthy tserver with idle + active backends, then relaunch the tserver ---------
for i in 1 2 3 4; do idle_client 3603; done
for i in 5 6; do ( $PSQL -c 'SELECT pg_sleep(120);' >"$W/active-$i.out" 2>&1 ) & done
t=$(wait_atleast 6 20); sleep 2; sample D "4 idle + 2 active sessions (+${t}s)"
res D states_before="$(probe_states)" pgprocs_before=$(npgprocs)
TS=$(tserver_pid); TS_CMDLINE=$(tr '\0' ' ' <"/proc/$TS/cmdline"); kill -9 "$TS"; log "D: SIGKILL tserver $TS"
sleep 1; sample D "+1s after tserver SIGKILL"
res D pgprocs_at_1s=$(npgprocs) probe_at_1s=$(nprobe)
t=$(wait_pgprocs_zero 20); sample D "all postgres processes gone? (+${t}s after the +1s sample)"
res D seconds_to_zero_pgprocs_after_1s=$t pgprocs_left=$(npgprocs) probe_left=$(nprobe)
pgprocs | sed 's/^/D: leftover postgres process: /'
close_clients 3603
relaunch_tserver D "$TS"
wait_ysql 120 && log "D: ysql ok after tserver relaunch, postmaster pid=$(postmaster_pid)" || log "D: ysql NOT ok after relaunch"
res D ysql_after_relaunch=$(ysql_ok && echo ok || echo fail) pgprocs_after_relaunch=$(npgprocs) probe_after_relaunch=$(nprobe)
res D tserver_cleanup_log="$(tslog_grep 'older postgres process|stale postgres lock file|PostgreSQL server running' 3 | sed -E 's/^[^]]*\] //' | cut -c1-140 | tr '\n' '|')"
rm -f "$W"/active-*.out

# ---- F: SIGKILL a frozen tserver while idle + blocked backends exist, then relaunch --------------
for i in 1 2 3; do idle_client 3604; done
for i in 4 5; do blocked_client SELECT $i; done
blocked_client INSERT 6
t=$(wait_atleast 6 20); sample F "3 idle + 3 querying sessions (+${t}s)"
TS=$(tserver_pid); TS_CMDLINE=$(tr '\0' ' ' <"/proc/$TS/cmdline"); kill -STOP "$TS"; log "F: SIGSTOP tserver $TS"
sleep 4; sample F "+4s: DocDB requests outstanding against the frozen tserver"
res F states_before="$(probe_states)" pgprocs_before=$(npgprocs)
kill -9 "$TS"; log "F: SIGKILL tserver $TS (while stopped)"
sleep 1; sample F "+1s after tserver SIGKILL"
res F pgprocs_at_1s=$(npgprocs) probe_at_1s=$(nprobe)
t=$(wait_pgprocs_zero 20); sample F "all postgres processes gone? (+${t}s after the +1s sample)"
res F seconds_to_zero_pgprocs_after_1s=$t pgprocs_left=$(npgprocs) probe_left=$(nprobe)
pgprocs | sed 's/^/F: leftover postgres process: /'
res F recalcitrant_sigkill_log=$(pglog_grep 'recalcitrant' 5 | wc -l | tr -d ' ')
close_clients 3604
relaunch_tserver F "$TS"
wait_ysql 120 && log "F: ysql ok after tserver relaunch, postmaster pid=$(postmaster_pid)" || log "F: ysql NOT ok after relaunch"
res F ysql_after_relaunch=$(ysql_ok && echo ok || echo fail) pgprocs_after_relaunch=$(npgprocs) probe_after_relaunch=$(nprobe)
res F client_errors="$(cat "$W"/blocked-*.out 2>/dev/null | sort | uniq -c | sed 's/^ *//' | cut -c1-120 | tr '\n' '|')"
log "== yb4577 probe v2 end; final postgres procs=$(npgprocs) probe backends=$(nprobe)"
res final pgprocs=$(npgprocs) probe_backends=$(nprobe)

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4577-run-matrix.txt`):

#!/bin/bash
# Host driver: run yb4577-probe.txt against one official image tag in a fresh podman container.
# usage: bash yb4577-run-matrix.txt <label> <image-tag>
# Output: /tmp/claude/yb4577/yb4577-cell-<label>.txt (full probe log incl. RESULT lines)
set -u
label=$1; tag=$2
IMG=docker.io/yugabytedb/yugabyte:$tag
C=yb4577-$label
OUT=/tmp/claude/yb4577/yb4577-cell-$label.txt
PROBE=/tmp/claude/yb4577/yb4577-probe.txt
podman rm -f "$C" >/dev/null 2>&1
{
  echo "== $(date -u +%FT%TZ) image=$IMG container=$C probe_sha256=$(shasum -a 256 "$PROBE" | cut -c1-64)"
  podman image inspect "$IMG" --format 'image digest={{.Digest}} id={{.Id}} arch={{.Architecture}}'
  podman run -d --init --name "$C" "$IMG" sleep infinity >/dev/null || { echo "run failed"; exit 1; }
  podman cp "$PROBE" "$C:/tmp/yb4577-probe.sh"
  podman exec "$C" bash -c '/home/yugabyte/bin/yb-tserver --version 2>&1 | head -1'
  podman exec "$C" bash -c '/home/yugabyte/bin/yugabyted start --base_dir=/tmp/ybd --advertise_address=127.0.0.1 2>&1 | grep -v -E "Slack|t-shirt" | tail -n 3'
  for i in $(seq 1 240); do
    podman exec "$C" bash -c '/home/yugabyte/bin/ysqlsh -h 127.0.0.1 -c "select 1" >/dev/null 2>&1' && { echo "ysql ready after ${i}s"; break; }
    sleep 1
  done
  podman exec "$C" bash /tmp/yb4577-probe.sh
  echo "== $(date -u +%FT%TZ) done; removing container"
} 2>&1 | tee "$OUT"
podman rm -f "$C" >/dev/null 2>&1

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4577-matrix-summary.txt`):

yugabyte/yugabyte-db#4577 -- runtime matrix summary (probe v2 sha256 7f82e2ef3b7ef6905640e23311113829e87e362571d95ab168467a488e2cae4b; run 2026-09-05 20:28-20:33 UTC; podman aarch64; yugabyted RF1 in one container per image)
Probe: /tmp/claude/yb4577/yb4577-probe.txt ; driver: /tmp/claude/yb4577/yb4577-run-matrix.txt ; raw logs: /tmp/claude/yb4577/yb4577-cell-<label>.txt
Reported signature (#4577): postgres backends left over / accumulating across a Jepsen run, most parked in the startup phase ('authentication'), some in SELECT/INSERT.
Phases: control(6 idle sessions open/close) | A(tserver SIGSTOP; 6 connects w/ 3s client timeout) | B(6 sessions blocked in SELECT/INSERT on frozen tserver; SIGINT postmaster)
        C(SIGKILL postmaster; 3 idle + 3 idle-in-transaction) | E(SIGKILL postmaster while 6 backends parked in startup phase, tserver frozen) | D(SIGKILL healthy tserver; 4 idle + 2 active) | F(SIGKILL frozen tserver; 3 idle + 3 blocked)

=== 2024.2: 2024.2.11.0-b36  image=docker.io/yugabytedb/yugabyte:2024.2.11.0-b36-aarch64
    version 2024.2.11.0 build 36 revision 04544fd94fb4e50d05994784f444b075a4a5dfdb build_type RELEASE built at 18 Aug 2026 00:58:07 UTC
    baseline pgprocs                                      6
    control  connected                                    6
    control  seconds_to_zero_after_client_close           1
    control  leftover                                     0
    A        label                                        authentication=6;
    A        startup_backends_exit_seconds_after_connect  9
    A        all_probe_backends_zero_while_frozen         yes
    A        leftover_after_cont                          0
    A        pglog_startup_exit_reason                    [684] FATAL:  Heartbeat RPC (request call id 1) to 127.0.0.1:9100 timed out after 9.000s
    B        blocked_states                               INSERT=3;SELECT=3;
    B        seconds_to_zero_after_sigint                 5
    B        leftover                                     0
    B        postmaster_alive                             no
    B        ysql_after_cont                              ok
    C        states_before                                idle=3;idle in transaction=3;
    C        orphans_at_1s                                0
    C        leftover_after_client_close                  0
    C        ysql_after                                   ok
    E        label                                        authentication=6;
    E        orphans_at_1s                                0
    E        orphans_at_3s                                0
    E        leftover_after_cont                          0
    E        ysql_after                                   ok
    D        states_before                                SELECT=2;idle=4;
    D        pgprocs_at_1s                                0
    D        seconds_to_zero_pgprocs_after_1s             0
    D        probe_left                                   0
    D        tserver_restarted_by                         yugabyted
    D        ysql_after_relaunch                          ok
    D        probe_after_relaunch                         0
    F        states_before                                INSERT=1;SELECT=2;idle=3;
    F        pgprocs_at_1s                                7
    F        probe_at_1s                                  3
    F        seconds_to_zero_pgprocs_after_1s             4
    F        probe_left                                   0
    F        recalcitrant_sigkill_log                     0
    F        ysql_after_relaunch                          ok
    F        probe_after_relaunch                         0
    final    pgprocs                                      5
    final    probe_backends                               0

=== 2025.1: 2025.1.4.0-b103  image=docker.io/yugabytedb/yugabyte:2025.1.4.0-b103-aarch64
    version 2025.1.4.0 build 103 revision 690ac4141e57d060fd834312f396643e8c41f260 build_type RELEASE built at 02 Apr 2026 20:43:14 UTC
    baseline pgprocs                                      6
    control  connected                                    6
    control  seconds_to_zero_after_client_close           1
    control  leftover                                     0
    A        label                                        initializing=6;
    A        startup_backends_exit_seconds_after_connect  9
    A        all_probe_backends_zero_while_frozen         yes
    A        leftover_after_cont                          0
    A        pglog_startup_exit_reason                    [703] FATAL:  Heartbeat RPC (request call id 1) to 127.0.0.1:9100 timed out after 9.000s
    B        blocked_states                               INSERT=3;SELECT=3;
    B        seconds_to_zero_after_sigint                 0
    B        leftover                                     0
    B        postmaster_alive                             yes
    B        ysql_after_cont                              ok
    C        states_before                                idle=3;idle in transaction=3;
    C        orphans_at_1s                                0
    C        leftover_after_client_close                  0
    C        ysql_after                                   ok
    E        label                                        initializing=6;
    E        orphans_at_1s                                0
    E        orphans_at_3s                                0
    E        leftover_after_cont                          0
    E        ysql_after                                   ok
    D        states_before                                SELECT=2;idle=4;
    D        pgprocs_at_1s                                0
    D        seconds_to_zero_pgprocs_after_1s             0
    D        probe_left                                   0
    D        tserver_restarted_by                         yugabyted
    D        ysql_after_relaunch                          ok
    D        probe_after_relaunch                         1
    F        states_before                                INSERT=1;SELECT=2;idle=3;
    F        pgprocs_at_1s                                6
    F        probe_at_1s                                  3
    F        seconds_to_zero_pgprocs_after_1s             timeout(last=5)
    F        probe_left                                   0
    F        recalcitrant_sigkill_log                     1
    F        ysql_after_relaunch                          ok
    F        probe_after_relaunch                         0
    final    pgprocs                                      5
    final    probe_backends                               0

=== 2025.2: 2025.2.6.0-b111  image=docker.io/yugabytedb/yugabyte:2025.2.6.0-b111-aarch64
    version 2025.2.6.0 build 111 revision ae22ef35ff65de63303467d95ac3e6d2382feba9 build_type RELEASE built at 02 Sep 2026 16:42:37 UTC
    baseline pgprocs                                      6
    control  connected                                    6
    control  seconds_to_zero_after_client_close           1
    control  leftover                                     0
    A        label                                        initializing=6;
    A        startup_backends_exit_seconds_after_connect  9
    A        all_probe_backends_zero_while_frozen         yes
    A        leftover_after_cont                          0
    A        pglog_startup_exit_reason                    [712] FATAL:  Heartbeat RPC (request call id 1) to 127.0.0.1:9100 timed out after 9.000s
    B        blocked_states                               INSERT=3;SELECT=3;
    B        seconds_to_zero_after_sigint                 1
    B        leftover                                     0
    B        postmaster_alive                             yes
    B        ysql_after_cont                              ok
    C        states_before                                idle=3;idle in transaction=3;
    C        orphans_at_1s                                0
    C        leftover_after_client_close                  0
    C        ysql_after                                   ok
    E        label                                        initializing=6;
    E        orphans_at_1s                                0
    E        orphans_at_3s                                0
    E        leftover_after_cont                          0
    E        ysql_after                                   ok
    D        states_before                                SELECT=2;idle=4;
    D        pgprocs_at_1s                                0
    D        seconds_to_zero_pgprocs_after_1s             0
    D        probe_left                                   0
    D        tserver_restarted_by                         yugabyted
    D        ysql_after_relaunch                          ok
    D        probe_after_relaunch                         0
    F        states_before                                INSERT=1;SELECT=2;idle=3;
    F        pgprocs_at_1s                                0
    F        probe_at_1s                                  0
    F        seconds_to_zero_pgprocs_after_1s             0
    F        probe_left                                   0
    F        recalcitrant_sigkill_log                     0
    F        ysql_after_relaunch                          ok
    F        probe_after_relaunch                         0
    final    pgprocs                                      5
    final    probe_backends                               0

=== 2026.1: 2026.1.1.1-b2  image=docker.io/yugabytedb/yugabyte:2026.1.1.1-b2-aarch64
    version 2026.1.1.1 build 2 revision 3b010e021f6027744bf9e016b17e108a2eb24e44 build_type RELEASE built at 14 Aug 2026 04:03:14 UTC
    baseline pgprocs                                      8
    control  connected                                    6
    control  seconds_to_zero_after_client_close           1
    control  leftover                                     0
    A        label                                        initializing=6;
    A        startup_backends_exit_seconds_after_connect  9
    A        all_probe_backends_zero_while_frozen         yes
    A        leftover_after_cont                          0
    A        pglog_startup_exit_reason                    [803] FATAL:  Heartbeat RPC (request call id 1) to 127.0.0.1:9100 timed out after 9.000s
    B        blocked_states                               INSERT=3;SELECT=3;
    B        seconds_to_zero_after_sigint                 1
    B        leftover                                     0
    B        postmaster_alive                             yes
    B        ysql_after_cont                              ok
    C        states_before                                idle=3;idle in transaction=3;
    C        orphans_at_1s                                0
    C        leftover_after_client_close                  0
    C        ysql_after                                   ok
    E        label                                        initializing=6;
    E        orphans_at_1s                                0
    E        orphans_at_3s                                0
    E        leftover_after_cont                          0
    E        ysql_after                                   ok
    D        states_before                                SELECT=2;idle=4;
    D        pgprocs_at_1s                                0
    D        seconds_to_zero_pgprocs_after_1s             0
    D        probe_left                                   0
    D        tserver_restarted_by                         yugabyted
    D        ysql_after_relaunch                          ok
    D        probe_after_relaunch                         0
    F        states_before                                INSERT=1;SELECT=2;idle=3;
    F        pgprocs_at_1s                                0
    F        probe_at_1s                                  0
    F        seconds_to_zero_pgprocs_after_1s             0
    F        probe_left                                   0
    F        recalcitrant_sigkill_log                     0
    F        ysql_after_relaunch                          ok
    F        probe_after_relaunch                         0
    final    pgprocs                                      5
    final    probe_backends                               0

=== master (preview): 2.25.2.0-b359  image=docker.io/yugabytedb/yugabyte:2.25.2.0-b359-aarch64
    version 2.25.2.0 build 359 revision 03bbc9f31eae5c3fecabd8f0afa5e99211e6080d build_type RELEASE built at 08 May 2025 19:05:35 UTC
    baseline pgprocs                                      5
    control  connected                                    6
    control  seconds_to_zero_after_client_close           1
    control  leftover                                     0
    A        label                                        initializing=6;
    A        startup_backends_exit_seconds_after_connect  15
    A        all_probe_backends_zero_while_frozen         yes
    A        leftover_after_cont                          0
    A        pglog_startup_exit_reason                    (no matching FATAL line captured)
    B        blocked_states                               INSERT=3;SELECT=3;
    B        seconds_to_zero_after_sigint                 timeout(last=6)
    B        leftover                                     6
    B        postmaster_alive                             yes
    B        ysql_after_cont                              ok
    C        states_before                                idle=3;idle in transaction=3;
    C        orphans_at_1s                                0
    C        leftover_after_client_close                  0
    C        ysql_after                                   ok
    E        label                                        initializing=6;
    E        orphans_at_1s                                0
    E        orphans_at_3s                                0
    E        leftover_after_cont                          0
    E        ysql_after                                   ok
    D        states_before                                SELECT=2;idle=4;
    D        pgprocs_at_1s                                0
    D        seconds_to_zero_pgprocs_after_1s             0
    D        probe_left                                   0
    D        tserver_restarted_by                         yugabyted
    D        ysql_after_relaunch                          ok
    D        probe_after_relaunch                         0
    F        states_before                                INSERT=1;SELECT=2;idle=3;
    F        pgprocs_at_1s                                0
    F        probe_at_1s                                  0
    F        seconds_to_zero_pgprocs_after_1s             0
    F        probe_left                                   0
    F        recalcitrant_sigkill_log                     0
    F        ysql_after_relaunch                          ok
    F        probe_after_relaunch                         0
    final    pgprocs                                      5
    final    probe_backends                               0

=== master (preview) supplementary: phase B with a 150s observation window (raw: /tmp/claude/yb4577/yb4577-cell-master-B-long.txt)
    RESULT B_long blocked_states=INSERT=3;SELECT=3;
    20:34:30 +1s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=10 postmaster_alive=yes
    20:35:29 +60s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=9 postmaster_alive=yes
    20:36:59 +150s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=9 postmaster_alive=yes
    RESULT B_long seconds_to_zero_after_sigint_frozen=not_within_150s(last=6)
    RESULT B_long pglog_tail=
    20:37:06 after SIGCONT: probe_backends=5 [INSERT=3;SELECT=2;] pgprocs=8 postmaster_alive=yes
    RESULT B_long ysql_after_cont=fail new_postmaster=217 pgprocs=8 probe_backends=5

Reading guide:
  orphans_at_1s=0 .............................. no backend survived its postmaster's SIGKILL (backend PR_SET_PDEATHSIG=SIGKILL: 2024.2 postmaster.c BackendStartup; PG15 lineage pmsignal.c PostmasterDeathSignalInit).
  startup_backends_exit_seconds_after_connect=9  backends parked in the startup phase on an unresponsive tserver exit at the 9s create-session heartbeat RPC bound (pg_client.cc PrepareHeartbeatController).
  pgprocs_at_1s=0 after a tserver SIGKILL ...... the whole postgres tree exited within 1s (tserver->postmaster PDEATHSIG: SIGINT fast shutdown; SIGQUIT on 2.25.2.0).
  2025.1 F 'timeout(last=5)' ................... the 5 processes are the NEW postmaster tree of the tserver that yugabyted restarted (probe_left=0, recalcitrant_sigkill_log=1); not residue.
  2025.1 D probe_after_relaunch=1 .............. sampling race with the probe's own 'select 1' health check backend; final probe_backends=0.
  master B 'timeout(last=6)' / B_long .......... 2.25.2.0-b359 predates #26758 (fast-shutdown SIGKILL escalation, master 0020131278, 2025-06-06): an EXTERNAL SIGINT to the postmaster while the tserver is unresponsive leaves the blocked backends attached to a postmaster stuck in shutdown (no orphans; no new connections accepted). The tserver's own paths on 2.25.2.0 use SIGQUIT (D/F: clean).

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4577-cell-master-B-long.txt`):

== 2026-09-05T20:34:07Z image=docker.io/yugabytedb/yugabyte:2.25.2.0-b359-aarch64 container=yb4577-master-blong

Starting yugabyted...
ysql ready after 1s
20:34:23 == yb4577 B-long start; version: version 2.25.2.0 build 359 revision 03bbc9f31eae5c3fecabd8f0afa5e99211e6080d build_type RELEASE built at 08 May 2025 19:05:35 UTC
20:34:25 SIGSTOP tserver 93 (postmaster 217)
20:34:29 +4s frozen: probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=11 postmaster_alive=yes
RESULT B_long blocked_states=INSERT=3;SELECT=3;
20:34:29 SIGINT postmaster 217 (fast shutdown)
20:34:30 +1s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=10 postmaster_alive=yes
20:34:34 +5s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=10 postmaster_alive=yes
20:34:39 +10s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=10 postmaster_alive=yes
20:34:49 +20s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=9 postmaster_alive=yes
20:34:59 +30s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=9 postmaster_alive=yes
20:35:14 +45s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=9 postmaster_alive=yes
20:35:29 +60s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=9 postmaster_alive=yes
20:35:44 +75s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=9 postmaster_alive=yes
20:35:59 +90s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=9 postmaster_alive=yes
20:36:14 +105s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=9 postmaster_alive=yes
20:36:29 +120s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=9 postmaster_alive=yes
20:36:44 +135s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=9 postmaster_alive=yes
20:36:59 +150s after SIGINT (tserver still frozen): probe_backends=6 [INSERT=3;SELECT=3;] pgprocs=9 postmaster_alive=yes
RESULT B_long seconds_to_zero_after_sigint_frozen=not_within_150s(last=6)
RESULT B_long pglog_tail=
20:36:59 SIGCONT tserver
20:37:06 after SIGCONT: probe_backends=5 [INSERT=3;SELECT=2;] pgprocs=8 postmaster_alive=yes
RESULT B_long ysql_after_cont=fail new_postmaster=217 pgprocs=8 probe_backends=5
RESULT B_long client_errors=1 |1                                                  ^|1 ERROR:  Shutdown connection|1 FATAL:  terminating connection due to administrator command|1 LINE 1: SELECT pg_sleep(3); SELECT count(*) FROM t;|
20:38:38 == yb4577 B-long end
== 2026-09-05T20:38:38Z done

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4577-cell-2024.2.txt`):

== 2026-09-05T20:28:58Z image=docker.io/yugabytedb/yugabyte:2024.2.11.0-b36-aarch64 container=yb4577-2024.2 probe_sha256=7f82e2ef3b7ef6905640e23311113829e87e362571d95ab168467a488e2cae4b
image digest=sha256:53c01c56315e245cc901cbc46e93f30fedb1f87833c7b1aa09d20fbeda27a288 id=e54b83a78b1677e6bc7f1d989f26a572de3b3a7d9204078d017e5040f7e24720 arch=arm64
version 2024.2.11.0 build 36 revision 04544fd94fb4e50d05994784f444b075a4a5dfdb build_type RELEASE built at 18 Aug 2026 00:58:07 UTC
🚀 YugabyteDB started successfully! To load a sample dataset, try 'yugabyted demo'.

Starting yugabyted...
ysql ready after 1s
20:29:17 == yb4577 probe v2 start; version: version 2024.2.11.0 build 36 revision 04544fd94fb4e50d05994784f444b075a4a5dfdb build_type RELEASE built at 18 Aug 2026 00:58:07 UTC
20:29:17 kernel: 7.0.12-201.fc44.aarch64; ps: ps from procps-ng 3.3.15
20:29:17 tserver pid=98 postmaster pid=241 (ppid of postmaster: 98)
20:29:17 tserver cmdline: /home/yugabyte/bin/yb-tserver --stop_on_parent_termination --undefok=stop_on_parent_termination --fs_data_dirs=/tmp/ybd/data --webserver_interface=127.0.0.1 --metrics_snapshotter_tserver_metrics_whitelist=handler_latency_yb_tserver_TabletServerService_Read_count,handler_latency_yb_tserver_TabletServerService_Write_count,handler_latency_yb_tserver_TabletServerService_Read_sum,handler_latency_yb_tserver_TabletServerService_Write_sum,disk_usage,cpu_usage,node_up --yb_num_shards_per_tserver=1 --ysql_num_shards_per_tserver=1 --placement_cloud=cloud1 --placement_region=datacenter1 --placement_zone=rack1 --use_memory_defaults_optimized_for_ysql=true --rpc_bind_addresses=127.0.0.1:9100 --server_broadcast_addresses=127.0.0.1:9100 --cql_proxy_bind_address=127.0.0.1:9042 --server_dump_info_path=/tmp/ybd/data/tserver-info --start_pgsql_proxy --pgsql_proxy_bind_address=127.0.0.1:5433 --tserver_enable_metrics_snapshotter=true --metrics_snapshotter_interval_ms=11000 --webserver_port=9000 --instance_uuid_override=5a90d08b80da43e3b6bf24c65a998fbd --start_redis_proxy=false --placement_uuid=4531d0e5-dd3d-4e44-be73-6101d95a707e --tserver_master_addrs=127.0.0.1:7100 
20:29:18 pre-existing (non-probe) client backends: postgres: yugabyte yugabyte 127.0.0.1(52754) idle|
20:29:18 all postgres procs at start: 6; probe backends: 0
RESULT baseline pgprocs=6 probe_backends=0
20:29:19 control: 6 idle sessions opened (+1s) probe_backends=6 [idle=6;] ppids=[241,] pgprocs=12
RESULT control connected=6 states=idle=6;
20:29:20 control: clients closed (+1s) probe_backends=0 [] ppids=[] pgprocs=6
RESULT control seconds_to_zero_after_client_close=1 leftover=0
20:29:20 A: SIGSTOP tserver 98
20:29:25 A: +5s (clients gave up at 3s) probe_backends=6 [authentication=6;] ppids=[241,] pgprocs=12
RESULT A startup_phase_backends_at_5s=6 label=authentication=6;
RESULT A client_result=6 ysqlsh: timeout expired|
20:29:29 A: startup-phase backends gone while tserver still frozen (+4s after the +5s sample) probe_backends=0 [] ppids=[] pgprocs=6
RESULT A startup_backends_exit_seconds_after_connect=9 still_frozen=yes
RESULT A all_probe_backends_zero_while_frozen=yes
20:29:29 A: SIGCONT tserver
20:29:29 A: after SIGCONT (+0s) probe_backends=0 [] ppids=[] pgprocs=6
RESULT A leftover_after_cont=0
20:29:29 A: ysql ok
RESULT A ysql_after_cont=ok
RESULT A pglog_startup_exit_reason=[684] FATAL:  Heartbeat RPC (request call id 1) to 127.0.0.1:9100 timed out after 9.000s|[682] FATAL:  Heartbeat RPC (request call id 1) to 127.0.0.1:9100 timed out after 9.000s|
20:29:30 B: 6 sessions connected, running pg_sleep(3) (+1s) probe_backends=6 [SELECT=6;] ppids=[241,] pgprocs=12
20:29:30 B: SIGSTOP tserver 98
20:29:34 B: +4s: pg_sleep done, DocDB requests outstanding against the frozen tserver probe_backends=6 [INSERT=3;SELECT=3;] ppids=[241,] pgprocs=12
RESULT B blocked_states=INSERT=3;SELECT=3; blocked_count=6
20:29:34 B: SIGINT postmaster 241 (fast shutdown)
20:29:39 B: after SIGINT (+5s, tserver still frozen) probe_backends=0 [] ppids=[] pgprocs=0
RESULT B seconds_to_zero_after_sigint=5 leftover=0 pgprocs_left=0 postmaster_alive=no
RESULT B recalcitrant_sigkill_log=0
20:29:39 B: SIGCONT tserver
20:29:41 B: ysql ok, postmaster pid=1289
RESULT B ysql_after_cont=ok new_postmaster=1289
RESULT B client_errors=6 	This probably means the server terminated abnormally|6 	before or while processing the request.|6 connection to server was lost|6 server closed the connection unexpectedly|
20:29:44 C: 6 sessions (+1s) probe_backends=6 [idle=3;idle in transaction=3;] ppids=[1289,] pgprocs=11
RESULT C states_before=idle=3;idle in transaction=3;
20:29:44 C: SIGKILL postmaster 1289
20:29:45 C: +1s after postmaster SIGKILL probe_backends=0 [] ppids=[] pgprocs=5
RESULT C orphans_at_1s=0 orphan_ppids= orphan_states=
20:29:45 C: clients still connected (+0s) probe_backends=0 [] ppids=[] pgprocs=5
RESULT C seconds_to_zero_clients_connected=0 leftover_clients_connected=0
20:29:45 C: clients closed (+0s) probe_backends=0 [] ppids=[] pgprocs=5
RESULT C leftover_after_client_close=0 seconds_to_zero_after_client_close=0
20:29:45 C: ysql ok, postmaster pid=1483
RESULT C ysql_after=ok new_postmaster=1483
RESULT C tserver_cleanup_log=Removing stale postgres lock file /tmp/.yb.127.0.0.1:5433/.s.PGSQL.5433.lock|Found older postgres process: 1289. It is not currently a PG process, so it will not be killed.|Found older postgres process: 1289. It is not currently a PG process, so it will not be killed.|
20:29:45 E: SIGSTOP tserver 98
20:29:49 E: +4s: startup-phase backends probe_backends=6 [authentication=6;] ppids=[1483,] pgprocs=11
RESULT E startup_phase_backends_before=6 label=authentication=6;
20:29:49 E: SIGKILL postmaster 1483
20:29:50 E: +1s after postmaster SIGKILL (tserver frozen) probe_backends=0 [] ppids=[] pgprocs=0
RESULT E orphans_at_1s=0 orphan_ppids= orphan_states=
20:29:52 E: +3s after postmaster SIGKILL (tserver frozen) probe_backends=0 [] ppids=[] pgprocs=0
RESULT E orphans_at_3s=0
20:29:52 E: waited for exit while tserver frozen (+0s after the +3s sample) probe_backends=0 [] ppids=[] pgprocs=0
RESULT E seconds_to_zero_while_frozen_after_3s=0
20:29:52 E: SIGCONT tserver
20:29:52 E: after SIGCONT (+0s) probe_backends=0 [] ppids=[] pgprocs=1
RESULT E leftover_after_cont=0
20:29:53 E: ysql ok, postmaster pid=2013
RESULT E ysql_after=ok new_postmaster=2013
RESULT E tserver_cleanup_log=Found older postgres process: 1289. It is not currently a PG process, so it will not be killed.|Found older postgres process: 1483. It is not currently a PG process, so it will not be killed.|Found older postgres process: 1483. It is not currently a PG process, so it will not be killed.|
20:29:56 D: 4 idle + 2 active sessions (+1s) probe_backends=6 [SELECT=2;idle=4;] ppids=[2013,] pgprocs=11
RESULT D states_before=SELECT=2;idle=4; pgprocs_before=11
20:29:56 D: SIGKILL tserver 98
20:29:57 D: +1s after tserver SIGKILL probe_backends=0 [] ppids=[] pgprocs=0
RESULT D pgprocs_at_1s=0 probe_at_1s=0
20:29:57 D: all postgres processes gone? (+0s after the +1s sample) probe_backends=0 [] ppids=[] pgprocs=0
RESULT D seconds_to_zero_pgprocs_after_1s=0 pgprocs_left=0 probe_left=0
20:29:59 D: tserver restarted by yugabyted (new pid 2356)
RESULT D tserver_restarted_by=yugabyted new_tserver=2356
20:29:59 D: ysql ok after tserver relaunch, postmaster pid=2417
RESULT D ysql_after_relaunch=ok pgprocs_after_relaunch=6 probe_after_relaunch=0
RESULT D tserver_cleanup_log=Found older postgres process: 1289. It is not currently a PG process, so it will not be killed.|Found older postgres process: 1483. It is not currently a PG process, so it will not be killed.|Found older postgres process: 1483. It is not currently a PG process, so it will not be killed.|
20:30:00 F: 3 idle + 3 querying sessions (+1s) probe_backends=6 [SELECT=3;idle=3;] ppids=[2417,] pgprocs=11
20:30:00 F: SIGSTOP tserver 2356
20:30:04 F: +4s: DocDB requests outstanding against the frozen tserver probe_backends=6 [INSERT=1;SELECT=2;idle=3;] ppids=[2417,] pgprocs=11
RESULT F states_before=INSERT=1;SELECT=2;idle=3; pgprocs_before=11
20:30:04 F: SIGKILL tserver 2356 (while stopped)
20:30:05 F: +1s after tserver SIGKILL probe_backends=3 [INSERT=1;SELECT=2;] ppids=[2417,] pgprocs=7
RESULT F pgprocs_at_1s=7 probe_at_1s=3
20:30:09 F: all postgres processes gone? (+4s after the +1s sample) probe_backends=0 [] ppids=[] pgprocs=0
RESULT F seconds_to_zero_pgprocs_after_1s=4 pgprocs_left=0 probe_left=0
RESULT F recalcitrant_sigkill_log=0
20:30:09 F: tserver restarted by yugabyted (new pid 2778)
RESULT F tserver_restarted_by=yugabyted new_tserver=2778
20:30:14 F: ysql ok after tserver relaunch, postmaster pid=2949
RESULT F ysql_after_relaunch=ok pgprocs_after_relaunch=5 probe_after_relaunch=0
RESULT F client_errors=3 	This probably means the server terminated abnormally|3 	before or while processing the request.|3 connection to server was lost|3 server closed the connection unexpectedly|
20:30:14 == yb4577 probe v2 end; final postgres procs=5 probe backends=0
RESULT final pgprocs=5 probe_backends=0
== 2026-09-05T20:30:14Z done; removing container

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu
